### PR TITLE
Apply schema inclusion/exclusion filtering to sql-jdbc drivers

### PR DESCRIPTION
**Pull request by jeff303**
Update sql-jdbc namespaces to handle schema inclusion/exclusion patterns when filtering schemas

Add new generic schema inclusion/exclusion test for sql-jdbc drivers that define the property

Update Snowflake and Redshift driver manifests to include schema filtering property

### CODE REVIEW

**Comment by jeff303** [comment]
- WORK IN PROGRESS

TODO:
- [ ] add new property to other JDBC drivers as mentioned in product doc
- [ ] figure out if non-working Snowflake "schema" property should be migrated to new pattern (i.e. #13126)

File: src/metabase/driver/sync.clj [hunk 1]
```diff
@@ -0,0 +1,51 @@
+(ns metabase.driver.sync
+  "General functions and utilities for sync operations across multiple drivers."
+  (:require [clojure.string :as str])
+  (:import java.util.regex.Pattern))
+
+(defn- schema-pattern->re-pattern ^Pattern [schema-pattern]
+  (re-pattern (-> (str/replace schema-pattern #"(^|[^\\\\])\*" "$1.*")
```
Comment (dpsutton) [inline src/metabase/driver/sync.clj:7]: this could use a comment to explain the regex.

File: src/metabase/driver/sync.clj [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+(defn include-schema?
+  "Returns true of the given `schema-name` should be included/synced, considering the given `inclusion-patterns` and
+  `exclusion-patterns`. Patterns are comma-separated, and can contain wildcard characters (`*`)."
+  {:added "0.42.0"}
+  [inclusion-patterns exclusion-patterns schema-name]
+  (let [filter-fn (schema-patterns->filter-fn inclusion-patterns exclusion-patterns)]
+    (filter-fn schema-name)))
+
+(defn db-details->schema-filter-patterns
+  "Given a `prop-nm` (which is expected to be a connection property of type `:schema-filters`), and a `database`
```
Comment (dpsutton) [inline src/metabase/driver/sync.clj:42]: is it ever not "schema-filters" right now? Is this open lookup in use now or just possible future use.

File: src/metabase/driver/sql_jdbc/sync/describe_database.clj [hunk 3]
```diff
@@ -89,32 +92,45 @@
...
     (eduction
      (comp (mapcat (fn [schema]
                      (db-tables driver metadata schema db-name-or-nil)))
            (filter (fn [{table-schema :schema, table-name :name}]
-                     (i/have-select-privilege? driver conn table-schema table-name))))
-     (i/syncable-schemas driver conn metadata))))
+                     (and
+                       (driver.s/include-schema? schema-inclusion-filters schema-exclusion-filters table-schema)
+                       (i/have-select-privilege? driver conn table-schema table-name)))))
+     (i/syncable-schemas driver conn metadata schema-inclusion-filters schema-exclusion-filters))))
```
Comment (dpsutton) [inline src/metabase/driver/sql_jdbc/sync/describe_database.clj:105]: This is a bit confusing to me. We filter `(and (driver.s/include-schema? inclusion-filters))` but also, the original seq is `(i/syncable-schemas inclusion-filters)`. It seems we are filtering twice. 

`i/syncable-schemas` for `:sql-jdbc`
```clojure
(defmethod i/syncable-schemas :sql-jdbc
  [driver _ metadata schema-inclusion-patterns schema-exclusion-patterns]
  (eduction (remove (set (i/excluded-schemas driver)))
            (filter (partial driver.s/include-schema? schema-inclusion-patterns schema-exclusion-patterns))
            (all-schemas metadata)))
```
which removes and includes.

For redshift, 
```clojure
(defmethod sql-jdbc.sync/syncable-schemas :redshift
  [driver conn metadata schema-inclusion-patterns schema-exclusion-patterns]
  (reducible-schemas-with-usage-permissions
   conn
   (eduction
    (remove (set (sql-jdbc.sync/excluded-schemas driver)))
    (filter (partial driver.s/include-schema? schema-inclusion-patterns schema-exclusion-patterns))
    (sync.describe-database/all-schemas metadata))))
```

So it seems that `i/syncable-schemas` runs the schemas through redshifts select privileges and then again through `:sql-jdbc`'s `i/have-select-privilege? :sql-jdbc` again, as well as the schema filtering. It seems there's a lot going on here, and lots of it twice.

This code is also hard to follow because of potemkin. Redshift implements `sql-jdbc.sync/syncable-schemas :redshift` but otherwise the default is `i/have-select-privilege? :sql-jdbc` because `metabase.driver.sql-jdbc.sync.interface` contains the actual defmethod whereas `metabase.driver.sql-jdbc.sync` potemkin import-vars this defmethod.

File: src/metabase/driver/util.clj [hunk 4]
```diff
@@ -198,6 +198,13 @@
       :placeholder (trs "You can use patterns like auth* to match multiple {0}" (str/lower-case disp-name))
       :visible-if {(keyword type-prop-nm) ["inclusion" "exclusion"]}}]))
 
+(defn find-schema-filters-prop
+  "Finds the first property of type `:schema-filters` for the given `driver` connection properties. Returns `nil`
+  if the driver has no property of that type."
+  [driver]
+  (first (filter (fn [conn-prop]
+                   (= :schema-filters (keyword (:type conn-prop))))
+           (driver/connection-properties driver))))
```
Comment (dpsutton) [inline src/metabase/driver/util.clj:207]: I'm looking at `driver/connection-properties` and i see implementations only for h2, mysql, and postgres. Presumably redshift goes through the postgres version because of the derivation but I'm surprised there's no default behavior or more implementations.

Comment (dpsutton) [reply src/metabase/driver/util.clj:207]: ah, nevermind i see the docstring about the lazy loading driver.

**Review by dpsutton — APPROVED** [review]
Approving but left some notes. I think the code is a bit confusing. Particularly

```clojure
(eduction
 (comp (mapcat (fn [schema]
                 (db-tables driver metadata schema db-name-or-nil)))
       (filter (fn [{table-schema :schema, table-name :name}]
                 (and
                  (driver.s/include-schema? schema-inclusion-filters schema-exclusion-filters table-schema)
                  (i/have-select-privilege? driver conn table-schema table-name)))))
 (i/syncable-schemas driver conn metadata schema-inclusion-filters schema-exclusion-filters))
```

Where we filter to the schemas we want, and then for each schema, get the tables in that schema, and for each table, filter that the schema is what we want. This makes me think that either a) `i/syncable-schemas` returns schemas that might not match the schema filters, b) `(db-tables ... schema)` might return tables that are not in the schema that was passed in, or c) we're filtering again for no reason.

I think we need a comment explaining why or we need to remove the duplicate filtering.

File: src/metabase/driver/sync.clj (hunk 1 above)
Comment (jeff303) [reply src/metabase/driver/sync.clj:7]: Added in next push.

File: src/metabase/driver/sync.clj (hunk 2 above)
Comment (jeff303) [reply src/metabase/driver/sync.clj:42]: No, it's always supposed to be of type `schema-filters`, which is the custom type I added in the previous BigQuery patch to perform this logic.  I just don't think it's needed to validate it at this point (since this function will only ever be called for that conn-prop type).  Does that make sense, or did I miss your point?

File: src/metabase/driver/sql_jdbc/sync/describe_database.clj (hunk 3 above)
Comment (jeff303) [reply src/metabase/driver/sql_jdbc/sync/describe_database.clj:111]: Yeah, there was definitely redundant filtering happening in the `:sql-jdbc` impl.  So I removed that.

Also, the `:redshift` impl actually did need the call to `include-schema?` since it was operating on `all-schemas`.  I changed it up to just delegate to the `:sql-jdbc` impl instead, and only apply the `reducible-schemas-with-usage-permissions` on top of those.

**Review by dpsutton — APPROVED** [review]

File: src/metabase/driver/sql_jdbc/sync/describe_database.clj (hunk 3 above)
Comment (camsaul) [reply src/metabase/driver/sql_jdbc/sync/describe_database.clj:111]: > This code is also hard to follow because of Potemkin.

`M-.` still jumps to the actual definition regardless of whether there's a Potemkin alias for it or not

File: src/metabase/driver/sql_jdbc/sync/interface.clj [hunk 5]
```diff
@@ -32,8 +35,14 @@
...
-  a sequence of all schema names from the JDBC database metadata and filter out any schemas in `excluded-schemas`."
-  {:added "0.39.0", :arglists '([driver ^java.sql.Connection connection ^java.sql.DatabaseMetaData metadata])}
+  a sequence of all schema names from the JDBC database metadata and filter out any schemas in `excluded-schemas`, along
+  with any that shouldn't be included based on the given inclusion and exclusion patterns (see the
+  `metabase.driver.sync` namespace for full explanation)."
+  {:added "0.39.0", :arglists '([driver
+                                 ^java.sql.Connection connection
+                                 ^java.sql.DatabaseMetaData metadata
+                                 ^String schema-inclusion-patterns
+                                 ^String schema-exclusion-patterns])}
```
Comment (camsaul) [inline src/metabase/driver/sql_jdbc/sync/interface.clj:45]: What happens to all the drivers that implemented the 3-arity version of this? 💀 

Kind of annoying but if we want to be responsible and developer-friendly we really ought to just make this a new method like we've done in similar situations in the past.

At any rate, changes also need to be mentioned in https://github.com/metabase/metabase/blob/master/docs/developers-guide/driver-changelog.md... since the file actually exists now it's probably best to do it as part of the PR so it can be reviewed together with the other changes

**Review by camsaul — CHANGES_REQUESTED** [review]
- Changing the signature of a driver multimethod will break any 3rd party drivers that implement that method. Since we want to be somewhat developer friendly it's probably better just to make a new method and deprecate the old one
- Now that we have a "what's new for driver authors in ____" page in the actual docs it would be nice to add changes to that doc as part of this PR
- You should invoke "parent methods" of multimethods using `((get-method ...) driver ...)` for reasons mentioned in https://github.com/metabase/metabase/pull/19651#discussion_r792007895

File: modules/drivers/redshift/src/metabase/driver/redshift.clj [hunk 6]
```diff
@@ -272,9 +271,7 @@
           reducible))))))
 
 (defmethod sql-jdbc.sync/syncable-schemas :redshift
-  [driver conn metadata]
+  [_ conn metadata schema-inclusion-patterns schema-exclusion-patterns]
```
Comment (camsaul) [inline modules/drivers/redshift/src/metabase/driver/redshift.clj:274]:
```suggestion
  [driver conn metadata schema-inclusion-patterns schema-exclusion-patterns]
```

File: modules/drivers/redshift/src/metabase/driver/redshift.clj [hunk 7]
```diff
@@ -272,9 +271,7 @@
...
   (reducible-schemas-with-usage-permissions
    conn
-   (eduction
-    (remove (set (sql-jdbc.sync/excluded-schemas driver)))
-    (sync.describe-database/all-schemas metadata))))
+   (sql-jdbc.sync/syncable-schemas :sql-jdbc conn metadata schema-inclusion-patterns schema-exclusion-patterns)))
```
Comment (camsaul) [inline modules/drivers/redshift/src/metabase/driver/redshift.clj:277]:
```suggestion
   ((get-method sql-jdbc.sync/syncable-schemas :sql-jdbc) driver conn metadata schema-inclusion-patterns schema-exclusion-patterns)))
```

Comment (camsaul) [reply modules/drivers/redshift/src/metabase/driver/redshift.clj:277]: You **definitely** should be invoking "parent methods" this way. Why? Because if `syncable-schemas` in turn calls some **other** multimethod, it's going to use the `driver` parameter you passed it for dispatch. If you pass it `:sql-jdbc`, then it's using the `:sql-jdbc` version of any methods it calls, even if there's a `:redshift`-specific override. So best practice here is to manually get the `:sql-jdbc` version of the method and then pass in the `driver` parameter you got.

Why not pass in `:redshift` directly instead of `driver`? Well, we don't know that someone isn't extending the Redshift driver in some way. This method might actually be getting invoked for some `:redshift-2` driver or something like that. So it's important to pass the actual `driver` we got, not what we think it is. (A good example of this would be the `:postgres` driver. The Postgres driver has to be explicit to pass around `driver` rather than `:postgres`, or else it would skip over Redshift's implementations of things)

Comment (jeff303) [reply modules/drivers/redshift/src/metabase/driver/redshift.clj:277]: Thanks, learned about this in the past but somehow forgot this time.

File: src/metabase/driver/sql_jdbc/sync/interface.clj (hunk 5 above)
Comment (jeff303) [reply src/metabase/driver/sql_jdbc/sync/interface.clj]: Fixed this, if you can have a look at the update.

File: modules/drivers/redshift/src/metabase/driver/redshift.clj (hunk 6 above)
Comment (jeff303) [reply modules/drivers/redshift/src/metabase/driver/redshift.clj]: Fixed in the update.

**Review by camsaul — APPROVED** [review]
Awesome. Changes look  good to me
